### PR TITLE
feat(remix-dev): make `entry.{client,server}.tsx` optional

### PR DESCRIPTION
**Pull request by mcansh**
made this before making a proposal like a dummy, but it's something we've talked about a bit internally

proposal here 👉 https://github.com/remix-run/remix/discussions/4601

follow PR to make remix config optional 🎉 

edit on the optional remix config.. turns out that already works as we check before we attempt to load it. https://github.com/remix-run/remix/blob/611c68bef0df137e79c99e1c6c111a0eabafd3f9/packages/remix-dev/config.ts#L321-L343

Signed-off-by: Logan McAnsh <logan@mcan.sh>

Closes: #

- [ ] Docs
- [x] Tests

Testing Strategy:

### CODE REVIEW

**Comment by MichaelDeBoey** [comment]
What about people using JS instead of TS?
Do they need to have this file in their repo if they never changed it?

Should we also remove these files from our templates/examples?

File: packages/remix-dev/cli/commands.ts [hunk 1]
```diff
@@ -347,3 +348,45 @@ function purgeAppRequireCache(buildPath: string) {
...
+  // 0. valid entry file
+  let entries = new Set([
+    "entry.server.tsx",
+    "entry.server.js",
+    "entry.server.jsx",
+    "entry.client.tsx",
+    "entry.client.js",
+    "entry.client.jsx",
+  ]);
+  if (!entries.has(entry)) {
```
Comment (MichaelDeBoey) [inline packages/remix-dev/cli/commands.ts:362]: `Array.includes` appeared to be faster

```suggestion
  if (!entries.includes(entry)) {
```

File: packages/remix-dev/cli/commands.ts [hunk 2]
```diff
@@ -347,3 +348,45 @@ function purgeAppRequireCache(buildPath: string) {
     }
   }
 }
+
+export async function generateEntry(remixRoot: string, entry: string) {
+  // 0. valid entry file
+  let entries = new Set([
```
Comment (MichaelDeBoey) [inline packages/remix-dev/cli/commands.ts:354]:
```suggestion
  let entries = [
```

File: packages/remix-dev/cli/commands.ts [hunk 3]
```diff
@@ -347,3 +348,45 @@ function purgeAppRequireCache(buildPath: string) {
...
+    "entry.server.tsx",
+    "entry.server.js",
+    "entry.server.jsx",
+    "entry.client.tsx",
+    "entry.client.js",
+    "entry.client.jsx",
+  ]);
```
Comment (MichaelDeBoey) [inline packages/remix-dev/cli/commands.ts:361]:
```suggestion
  ];
```

File: packages/remix-dev/cli/commands.ts [hunk 4]
```diff
@@ -347,3 +348,45 @@ function purgeAppRequireCache(buildPath: string) {
...
+  if (!entries.has(entry)) {
+    console.log(
+      colors.error(
+        `Invalid entry file. Valid entry files are ${Array.from(entries).join(
```
Comment (MichaelDeBoey) [inline packages/remix-dev/cli/commands.ts:365]:
```suggestion
        `Invalid entry file. Valid entry files are ${entries.join(
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]
I think we should also run the JS migration on the file when people choose the `js`/`jsx` extension in the generator command?

**Comment by mcansh** [comment]
> I think we should also run the JS migration on the file when people choose the `js`/`jsx` extension in the generator command?

definitely, this is still very much WIP

File: packages/remix-dev/cli/commands.ts [hunk 5]
```diff
@@ -234,3 +237,103 @@ export async function codemod(
...
+      contents = contents.replace(/@remix-run\/node/g, "@remix-run/cloudflare");
+    } else if (deps["@remix-run/deno"]) {
+      contents = contents.replace(/@remix-run\/node/g, "@remix-run/deno");
+    }
+  }
+
+  // 3. if entry is js/jsx, convert to js
+  // otherwise, copy the entry file from the defaults
+  if (/\.jsx?$/.test(entry)) {
+    contents = contents ||= await fse.readFile(inputFile, "utf-8");
```
Comment (MichaelDeBoey) [inline packages/remix-dev/cli/commands.ts:319]: I think you meant

```suggestion
    contents ||= await fse.readFile(inputFile, "utf-8");
```

Comment (mcansh) [reply packages/remix-dev/cli/commands.ts:319]: yes, yes i did 🤪 - gonna need to also update PassThrough in deno/cf too, so may go with something different anyways

**Comment by mcansh** [comment]
> What about people using JS instead of TS? 

if they "eject" we'll transform it to js

> Do they need to have this file in their repo if they never changed it?

doesn't hurt

> Should we also remove these files from our templates/examples?

that's up for debate, templates would get any update faster than it being bundled in the CLI (unless we hit GitHub to get the file and use the bundled one if they're offline)

File: packages/remix-dev/config/defaults/entry.client.tsx [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import * as React from "react";
```
Comment (MichaelDeBoey) [inline packages/remix-dev/config/defaults/entry.client.tsx:1]: I don't think this is necessary?

**Review by MichaelDeBoey — APPROVED** [review]

File: packages/remix-dev/config/defaults/entry.client.tsx [hunk 7]
```diff
@@ -13,10 +13,10 @@ function hydrate() {
...
-if (typeof requestIdleCallback === "function") {
-  requestIdleCallback(hydrate);
+if (window.requestIdleCallback) {
+  window.requestIdleCallback(hydrate);
 } else {
   // Safari doesn't support requestIdleCallback
   // https://caniuse.com/requestidlecallback
-  setTimeout(hydrate, 1);
+  window.setTimeout(hydrate, 1);
 }
```
Comment (MichaelDeBoey) [inline packages/remix-dev/config/defaults/entry.client.tsx:22]: I think we should keep these to check if global `requestIdleCallback` is a function (see #4998)

```suggestion
if (typeof requestIdleCallback === "function") {
  requestIdleCallback(hydrate);
} else {
  // Safari doesn't support requestIdleCallback
  // https://caniuse.com/requestidlecallback
  setTimeout(hydrate, 1);
}
```

File: packages/remix-dev/config/defaults/entry.server.cloudflare.tsx [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import { type EntryContext } from "@remix-run/cloudflare";
```
Comment (MichaelDeBoey) [inline packages/remix-dev/config/defaults/entry.server.cloudflare.tsx:1]:
```suggestion
import type { EntryContext } from "@remix-run/cloudflare";
```

File: packages/remix-dev/config/defaults/entry.server.deno.tsx [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import { type EntryContext } from "@remix-run/cloudflare";
```
Comment (MichaelDeBoey) [inline packages/remix-dev/config/defaults/entry.server.deno.tsx:1]:
```suggestion
import type { EntryContext } from "@remix-run/deno";
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

File: templates/cloudflare-pages/package.json [hunk 10]
```diff
@@ -8,7 +8,7 @@
     "dev": "remix build && run-p \"dev:*\"",
     "start": "cross-env NODE_ENV=production npm run wrangler",
     "typecheck": "tsc -b",
-    "wrangler": "wrangler pages dev ./public"
+    "wrangler": "wrangler pages dev ./public --compatibility-flags streams_enable_constructors"
```
Comment (MichaelDeBoey) [inline templates/cloudflare-pages/package.json:11]: I think putting this in `wrangler.toml` would be a better idea, as that makes all config is in 1 place (just like @lili21 did in #4502)

File: packages/remix-dev/config/defaults/entry.server.deno.tsx [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+export default async function handleRequest(
+  request: Request,
+  responseStatusCode: number,
+  responseHeaders: Headers,
+  remixContext: EntryContext
+) {
+  const body = await renderToReadableStream(
+    <RemixServer context={remixContext} url={request.url} />,
+    {
```
Comment (mcansh) [inline packages/remix-dev/config/defaults/entry.server.deno.tsx:14]: add `signal: request.signal` to allow aborting

File: packages/remix-dev/cli/commands.ts [hunk 12]
```diff
@@ -246,3 +249,107 @@ export async function codemod(
...
+      `Entry file ${entry} created at ${path.relative(remixRoot, outputFile)}.`
+    )
+  );
+
+  return process.exit(0);
+}
+
+async function checkForEntry(remixRoot: string, entries: string[]) {
+  for (let entryToCheck of entries) {
+    let entryPath = path.resolve(remixRoot, "app", entryToCheck);
```
Comment (mcansh) [inline packages/remix-dev/cli/commands.ts:332]: use `appDirectory` from config

File: packages/remix-dev/cli/commands.ts [hunk 13]
```diff
@@ -246,3 +249,107 @@ export async function codemod(
...
+    let entryExists = await fse.pathExists(entryPath);
+    if (entryExists) {
+      console.error(
+        colors.error(
+          `Entry file ${path.relative(remixRoot, entryToCheck)} already exists.`
+        )
+      );
+      return process.exit(1);
```
Comment (mcansh) [inline packages/remix-dev/cli/commands.ts:340]: don't exit, just warn

File: packages/remix-dev/compiler/plugins/cssBundleEntryModulePlugin.ts [hunk 14]
```diff
@@ -27,7 +27,7 @@ export function cssBundleEntryModulePlugin(config: RemixConfig): Plugin {
           contents: [
             // These need to be exports to avoid tree shaking
             `export * as entryClient from ${JSON.stringify(
-              `./${config.entryClientFile}`
+              config.entryClientFilePath
```
Comment (mcansh) [inline packages/remix-dev/compiler/plugins/cssBundleEntryModulePlugin.ts:30]: keep it relative to make output hash deterministic

File: packages/remix-dev/config/defaults/entry.server.cloudflare.tsx [hunk 15]
```diff
@@ -0,0 +1,32 @@
+import type { EntryContext } from "@remix-run/cloudflare";
+import { RemixServer } from "@remix-run/react";
+import isbot from "isbot";
```
Comment (mcansh) [inline packages/remix-dev/config/defaults/entry.server.cloudflare.tsx:3]: we should also check and install `isbot` if it's not already, then log that they should commit the change

File: templates/cloudflare-pages/app/entry.server.tsx [hunk 16]
```diff
@@ -1,21 +1,30 @@
...
+  const stream = await renderToReadableStream(
+    <RemixServer context={remixContext} url={request.url} />,
+    {
+      signal: request.signal,
+      onError(error: unknown) {
+        didError = true;
+        console.error(error);
+      },
+    }
   );
```
Comment (jacob-ebey) [inline templates/cloudflare-pages/app/entry.server.tsx:23]: need an "isbot" check.

File: templates/deno/app/entry.server.tsx [hunk 17]
```diff
@@ -1,22 +1,30 @@
...
+  const stream = await renderToReadableStream(
     <RemixServer context={remixContext} url={request.url} />,
+    {
+      signal: request.signal,
+      onError(error: unknown) {
+        didError = true;
+        console.error(error);
+      },
+    }
   );
```
Comment (jacob-ebey) [inline templates/deno/app/entry.server.tsx:23]: Also needs "isbot".

**Review by jacob-ebey — APPROVED** [review]

**Pull request merged**